clients/js/test/helps/lifecycle.test.ts

```diff
@@ -226,22 +306,446 @@ test('it can detect burnable when frozen with permanent collection burn delegate
+      },
+    },
+  }).sendAndConfirm(umi);
+
+  const asset = await createAsset(umi, {
+    owner,
+    plugins: [
+      {
+        type: 'Oracle',
+        baseAddress: oracle.publicKey,
+        resultsOffset: {
+          type: 'Anchor',
+        },
+        lifecycleChecks: {
+          transfer: [CheckResult.CAN_REJECT],
+        },
+      },
+      {
+        type: 'Oracle',
+        baseAddress: oracle2.publicKey,
+        resultsOffset: {
+          type: 'Anchor',
+        },
+        lifecycleChecks: {
+          update: [CheckResult.CAN_REJECT],
+        },
+      },
+    ],
+  });
+
+  t.is(
+    await validateTransfer(umi, {
+      authority: owner.publicKey,
+      asset,
+    }),
+    null
+  );
+});
+
+test('it can validate non-transferrable asset with oracle with recipient seed', async (t) => {
+  const umi = await createUmi();
+  const owner = generateSigner(umi);
+  const seedPubkey = generateSigner(umi).publicKey;
+  const newOwner = generateSigner(umi);
+
+  const oraclePlugin: OracleInitInfoArgs = {
+    type: 'Oracle',
+    resultsOffset: {
+      type: 'Anchor',
+    },
+    lifecycleChecks: {
+      transfer: [CheckResult.CAN_REJECT],
+    },
+    baseAddress: MPL_CORE_ORACLE_EXAMPLE_PROGRAM_ID,
+    pda: {
+      type: 'CustomPda',
+      seeds: [
+        { type: 'Collection' },
+        { type: 'Owner' },
+        { type: 'Recipient' },
+        { type: 'Asset' },
+        { type: 'Address', pubkey: seedPubkey },
+        {
+          type: 'Bytes',
+          bytes: Buffer.from('example-seed-bytes', 'utf8'),
+        },
+      ],
+    },
+  };
+
+  const { asset, collection } = await createAssetWithCollection(
+    umi,
+    {
+      owner,
+      plugins: [oraclePlugin],
+    },
+    {}
+  );
+
+  const account = findOracleAccount(umi, oraclePlugin, {
+    collection: collection.publicKey,
+    owner: owner.publicKey,
+    recipient: newOwner.publicKey,
+    asset: asset.publicKey,
+  });
+
+  // write to example program oracle account
+  await customPdaAllSeedsInit(umi, {
+    account,
+    signer: umi.identity,
+    payer: umi.identity,
+    args: {
+      collection: collection.publicKey,
+      owner: owner.publicKey,
+      recipient: newOwner.publicKey,
+      asset: asset.publicKey,
+      address: seedPubkey,
+      bytes: Buffer.from('example-seed-bytes', 'utf8'),
+      oracleData: {
+        __kind: 'V1',
+        create: ExternalValidationResult.Pass,
+        update: ExternalValidationResult.Pass,
+        transfer: ExternalValidationResult.Rejected,
+        burn: ExternalValidationResult.Pass,
+      },
+    },
+  }).sendAndConfirm(umi);
+
+  t.is(
+    await validateTransfer(umi, {
+      authority: owner.publicKey,
+      asset,
+      recipient: newOwner.publicKey,
+      collection,
+    }),
+    'Unable to transfer: oracle validation failed.'
+  );
+});
+
+test('it can validate and skip transferrable asset with oracle with recipient seed if missing recipient', async (t) => {
```

Comment:
I don't fully understand this test.  Is this test just saying it validates because it doesn't yet know the recipient?

And that is also why we don't need to actually setup the Oracle because we aren't actually checking it when recipient omitted from `validateTransfer`?

Reply:
so this is because in most UIs, you click the transfer button first before specifying the recipient, maybe the logic shouldn't be implicit and an option to the validate method?

Reply:
I looked at this and thought that's how it worked but needed to ask to be sure.  Therefore I think making `recipient` mandory but allow `null` could be clearer.

But i'm also not used to all of the TS conventions either, so it fine with me either way on your best judgement as the actual TS expert.